File: definitions/core/stopwatch.lua
---@meta

---@class Stopwatch
---@field elapsed number
---@field new fun(): Stopwatch
Stopwatch = {}

---@return Stopwatch
function Stopwatch.new() end

---@overload fun(e: Entity): Stopwatch
---@param e Entity
---@param sw Stopwatch
---@return Stopwatch
function Stopwatch.set(e, sw) end

---@param e Entity
---@return Stopwatch
function Stopwatch.get(e) end

---@param e Entity
---@return boolean
function Stopwatch.has(e) end

---@param e Entity
function Stopwatch.remove(e) end

---@return boolean
function Stopwatch:is_paused() end

---@return Stopwatch
function Stopwatch:reset() end

---@overload fun(self: Stopwatch, dt: number): Stopwatch
---@return Stopwatch
function Stopwatch:tick() end

---@return Stopwatch
function Stopwatch:pause() end

---@return Stopwatch
function Stopwatch:unpause() end

File: sources/core/src/ecs/stopwatch.cpp
#include <glint/core/ecs/stopwatch.hpp>

#include <raylib-cpp/raylib.hpp>

namespace glint::core {

auto StopwatchComp::elapsed(this const StopwatchComp& self) -> float {
    return self.m_elapsed;
}

[[nodiscard]]
auto StopwatchComp::is_paused(this const StopwatchComp& self) -> bool {
    return self.m_is_paused;
}

auto StopwatchComp::reset(this StopwatchComp& self) -> StopwatchComp& {
    self.m_elapsed = 0.0f;
    return self;
}

auto StopwatchComp::set_elapsed(this StopwatchComp& self, float elapsed) -> StopwatchComp& {
    self.m_elapsed = elapsed;
    return self;
}

auto StopwatchComp::tick(this StopwatchComp& self) -> StopwatchComp& {
    if (!self.m_is_paused) self.m_elapsed += GetFrameTime();
    return self;
}

auto StopwatchComp::tick(this StopwatchComp& self, float dt) -> StopwatchComp& {
    if (!self.m_is_paused) self.m_elapsed += dt;
    return self;
}

auto StopwatchComp::pause(this StopwatchComp& self) -> StopwatchComp& {
    self.m_is_paused = true;
    return self;
}

auto StopwatchComp::unpause(this StopwatchComp& self) -> StopwatchComp& {
    self.m_is_paused = false;
    return self;
}

} // namespace glint::core
